### ISSUE TITLE
Lint request: Prevent lifetimes that are only used for function output

### ISSUE THREAD

**Issue by xfix** [Manishearth/rust-clippy#1874]
The following:

```rust
fn lifetime<'a>() -> &'a str {
    "Why are you doing this"
}
```

Can be more clearly written as this. Also, using `'static` lifetime makes it clear that it was intended.

```rust
fn lifetime() -> &'static str {
    "Why are you doing this"
}
```

**Comment by clarcharr** [comment]
In the standard library the opposite is true; people tend to prefer the former style to the latter.

**Comment by oli-obk** [comment]
Yea, the static lifetime might mess with some inference around multiple lifetimes. If this lint is deemed too controversial, it can always become a restriction lint or allow-by-default

**Comment by clarcharr** [comment]
I should also add that this only works for `'static` types; for non-`'static` types you have to use the former style to get the same effect.

**Comment by xfix** [comment]
Oh, there is a difference. What about making this a case only for public non-`unsafe` functions where this is definitely a mistake.

**Comment by felix91gr** [comment]
Where are we with regards to this lint?

> What about making this a case only for public non-unsafe functions where this is definitely a mistake.

If this is what we want, I reckon I can probably implement it

**Comment by oli-obk** [comment]
The only way we can implement it without false positives is to check that the return type "has a `'static`" bound, which I'm not sure how to check reliably. I'm not sure if the complexity required for this lint is worth it considering the minor improvement in API that it brings.

**Comment by felix91gr** [comment]
Should there be a `needs-discussion` label in this issue then, @oli-obk? Maybe we can close this one if the complexity is indeed too much for the improvement it brings :)

**Comment by clarfon** [comment]
I personally think that a means to not have both `-> &'b T` and `<'a: 'b> -> &'a T` (the static case is just `'b: 'static`, which would make the `'a: 'static` bound redundant). The only cases I can think of where you can't explicitly name `'b` is when you're creating a type inside a block, which should be relatively easy to check for and bail out of linting.

Dunno if it's a good idea to toss this yet, I just think that it's not something that is high on the priority list.

**Comment by camsteffen** [comment]
Similar to #305
